ipaserver/install/ipa_cert_fix.py

```diff
@@ -0,0 +1,244 @@
+    'prevent the normal operation of FreeIPA.  It should ONLY be used',
+    'in such scenarios, and backup of the system, especially certificates',
+    'and keys, is STRONGLY RECOMMENDED.',
+    '',
+])
+
+logger = logging.getLogger(__name__)
+
+
+class IPACertType(Enum):
+    IPARA = "IPA RA"
+    HTTPS = "Apache HTTPS"
+    LDAPS = "LDAP"
+    KDC = "KDC"
+
+
+class IPACertFix(AdminTool):
+    command_name = "ipa-cert-fix"
+    usage = "%prog"
+    description = "Renew expired certificates."
+
+    def validate_options(self):
+        super(IPACertFix, self).validate_options(needs_root=True)
+
+    def run(self):
+        api.bootstrap(in_server=True, confdir=paths.ETC_IPA)
+        api.finalize()
+        api.Backend.ldap2.connect()  # ensure DS is up
+
+        now = datetime.datetime.now() + datetime.timedelta(weeks=2)
+        certs, extra_certs = expired_certs(now)
+
+        if not certs and not extra_certs:
+            print("Nothing to do.")
+            return 0
+
+        print(msg)
+
+        print_intentions(certs, extra_certs)
+
+        response = user_input('Enter "Yes" to proceed')
+        if response != 'Yes':
+            print("Not proceeding.")
+            return 0
+        print("Proceeding.")
+
+        run_cert_fix(certs, extra_certs)
+
+        replicate_dogtag_certs(certs)
+        install_ipa_certs(extra_certs)
+
+        if any(x != 'sslserver' for x in certs) \
+                or any(x[0] is IPACertType.IPARA for x in extra_certs):
+            # we renewed a "shared" certificate, therefore we must
+            # become the renewal master
+            print("Becoming renewal master.")
+            cainstance.CAInstance().set_renewal_master()
+
+        subprocess.check_call(['ipactl', 'restart'])
+
+
+def expired_certs(now):
+    return expired_dogtag_certs(now), expired_ipa_certs(now)
+
+
+def expired_dogtag_certs(now):
+    """
+    Determine which Dogtag certs are expired, or close to expiry.
+
+    Return a list of (cert_id, cert) pairs.
+
+    """
+    certs = []
+    db = NSSDatabase(nssdir=paths.PKI_TOMCAT_ALIAS_DIR)
+
+    for certid, nickname in [
+        ('sslserver', 'Server-Cert cert-pki-ca'),
+        ('subsystem', 'subsystemCert cert-pki-ca'),
+        ('ca_ocsp_signing', 'ocspSigningCert cert-pki-ca'),
+        ('ca_audit_signing', 'auditSigningCert cert-pki-ca'),
+    ]:
+        cert = db.get_cert(nickname)
+        if cert.not_valid_after <= now:
+            certs.append((certid, cert))
+
+    return certs
+
+
+def expired_ipa_certs(now):
+    """
+    Determine which IPA certs are expired, or close to expiry.
+
+    Return a list of (IPACertType, cert) pairs.
+
+    """
+    certs = []
+
+    # IPA RA
+    cert = x509.load_certificate_from_file(paths.RA_AGENT_PEM)
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.IPARA, cert))
+
+    # Apache HTTPD
+    db = NSSDatabase(nssdir=paths.HTTPD_ALIAS_DIR)
```

Comment:
```
>>> from ipapython.certdb import NSSDatabase
>>> from ipaplatform.paths import paths
>>> db = NSSDatabase(nssdir=paths.HTTPD_ALIAS_DIR)
>>> cert = db.get_cert('Server-Cert')
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/ipapython/certdb.py", line 829, in get_cert
    result = self.run_certutil(args, capture_output=True)
  File "/usr/lib/python3.7/site-packages/ipapython/certdb.py", line 328, in run_certutil
    return ipautil.run(new_args, stdin, cwd=self.secdir, **kwargs)
  File "/usr/lib/python3.7/site-packages/ipapython/ipautil.py", line 574, in run
    p.returncode, arg_string, output_log, error_log
ipapython.ipautil.CalledProcessError: CalledProcessError(Command ['/usr/bin/certutil', '-d', 'sql:/etc/httpd/alias', '-L', '-n', 'Server-Cert', '-a', '-f', '/etc/httpd/alias/pwdfile.txt'] returned non-zero exit status 255: 'certutil: function failed: SEC_ERROR_BAD_DATABASE: security library: bad database.\n')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/site-packages/ipapython/certdb.py", line 831, in get_cert
    raise RuntimeError("Failed to get %s" % nickname)
RuntimeError: Failed to get Server-Cert
```

Recent IPA versions use ``/var/lib/ipa/certs/httpd.crt`` and ``/var/lib/ipa/private/httpd.key``.

Reply:
This PR is for the `ipa-4-6` branch.  I will have for "forwardport" it to newer branches, which will include some changes to certificate locations.  But because reasons, the focus is on `ipa-4-6` first :)